src/cmap/background.rs

```diff
@@ -1,71 +1,78 @@
-    ensure_min_connections_in_pool(&pool);
+    pool.remove_perished_connections().await;
+    pool.ensure_min_connections().await;
 }
 
-/// Iterate over the connections and remove any that are stale or idle.
-fn remove_perished_connections_from_pool(pool: &ConnectionPool) {
-    let mut connections = pool.inner.connections.write().unwrap();
-    let mut i = 0;
+impl ConnectionPool {
+    /// Iterate over the connections and remove any that are stale or idle.
+    async fn remove_perished_connections(&self) {
+        let mut connection_manager = self.inner.connection_manager.lock().await;
 
-    while i < connections.len() {
-        if connections[i].is_stale(pool.inner.generation.load(Ordering::SeqCst)) {
-            pool.close_connection(connections.remove(i), ConnectionClosedReason::Stale);
-        } else if connections[i].is_idle(pool.inner.max_idle_time) {
-            pool.close_connection(connections.remove(i), ConnectionClosedReason::Idle);
-        } else {
-            i += 1;
+        let mut i = 0;
+        while i < connection_manager.checked_in_connections.len() {
+            if connection_manager.checked_in_connections[i].is_stale(connection_manager.generation) {
+                let connection = connection_manager.checked_in_connections.remove(i);
+                connection_manager.close_connection(connection, ConnectionClosedReason::Stale);
+            } else if connection_manager.checked_in_connections[i].is_idle(self.inner.max_idle_time) {
+                let connection = connection_manager.checked_in_connections.remove(i);
+                connection_manager.close_connection(connection, ConnectionClosedReason::Idle);
+            } else {
+                i += 1;
+            }
         }
     }
-}
-
-/// Add connections until the min pool size it met. We explicitly release the lock at the end of
-/// each iteration and acquire it again during the next one to ensure that the this method doesn't
-/// block other threads from acquiring connections.
-fn ensure_min_connections_in_pool(pool: &ConnectionPool) {
-    if let Some(min_pool_size) = pool.inner.min_pool_size {
-        loop {
-            let mut connections = pool.inner.connections.write().unwrap();
 
-            if pool.inner.total_connection_count.load(Ordering::SeqCst) < min_pool_size {
-                match pool.create_connection(false) {
-                    Ok(connection) => connections.push(connection),
-                    e @ Err(_) => {
-                        // Since we had to clear the pool, we return early from this function and
-                        // put the background thread back to sleep. Next time it wakes up, the
-                        // stale connections will be closed, and the thread can try to create new
-                        // ones after that.
-                        return;
+    /// Add connections until the min pool size it met. We explicitly release the lock at the end of
```

Comment:
Kind of weird that git doesn't realize that this is really similar to the section on the left just because of an extra closing brace